src/Microsoft.SqlTools.Authentication/MSALEncryptedCacheHelper.cs

```diff
@@ -0,0 +1,261 @@
+
+            this._ivKeyReadCallback = callback;
+
+            this.fillIvKeyIfNeeded();
+        }
+
+        /// <summary>
+        /// Registers <paramref name="tokenCache"/> before and after access methods that are fired on cache access.
+        /// </summary>
+        /// <param name="tokenCache">Access token cache from MSAL.NET</param>
+        /// <exception cref="ArgumentNullException">When token cache is not provided.</exception>
+        public void RegisterCache(ITokenCache tokenCache)
+        {
+            if (tokenCache == null)
+            {
+                throw new ArgumentNullException(nameof(tokenCache));
+            }
+
+            Logger.Information($"Registering MSAL token cache with encrypted file storage");
+
+            // If the token cache was already registered, this operation does nothing
+            tokenCache.SetBeforeAccess(BeforeAccessNotification);
+            tokenCache.SetAfterAccess(AfterAccessNotification);
+        }
+
+        #endregion
+
+        #region Private Methods
+
+        private void fillIvKeyIfNeeded()
+        {
+            if (this._key == null || this._iv == null)
+            {
+                this._ivKeyReadCallback(out string key, out string iv);
+
+                if (key != null)
+                {
+                    this._key = Encoding.Unicode.GetBytes(key);
+                }
+
+                if (iv != null)
+                {
+                    this._iv = Encoding.Unicode.GetBytes(iv);
+                }
+
+                Logger.Verbose($"Received IV and Key from callback");
+            }
+        }
+
+        /// <summary>
+        /// Triggered after cache is accessed, <paramref name="args"/> provides updated cache data that
+        /// needs to be updated in File Storage. We encrypt cache data here and store it in file system.
+        /// </summary>
+        /// <param name="args">Access token cache notification arguments.</param>
+        private void AfterAccessNotification(TokenCacheNotificationArgs args)
+        {
+            try
+            {
+                Logger.Verbose($"After access");
+                byte[]? data = null;
+                // if the access operation resulted in a cache update
+                if (args.HasStateChanged)
+                {
+                    Logger.Verbose($"After access, cache in memory HasChanged");
+                    try
+                    {
+                        data = args.TokenCache.SerializeMsalV3();
+                    }
+                    catch (Exception e)
+                    {
+                        Logger.Error($"An exception was encountered while serializing the {nameof(MsalCacheHelper)} : {e}");
+                        Logger.Error($"No data found in the store, clearing the cache in memory.");
+
+                        // The cache is corrupt clear it out
+                        this._cacheStorage.Clear(ignoreExceptions: true);
```

Comment:
should we propagate failures here (or at least log them), or they're safe to ignore in this case for some reason?

Reply:
These are safe to ignore, as we also do the same from NodeJS layer when cache gets corrupted/intruded.